### PR TITLE
[Feature] MO AI - Vector Embedding Support

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [X] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:
#10632 
#10633 

## What this PR does / why we need it:

As a part of [AI initiatives](https://github.com/orgs/matrixorigin/projects/22), we are introducing vector embedding support to MatrixOne.

## Operators/Functions supported

1. Elementwise Binary Operators : `+`, `-`,` *`,` /`, `=`,`!=`, `>`,` >=` , `<`, `<=`
2. Elementwise Unary Functions: `sqrt`, `abs`, `cast`
3. Auto casting:
    - `vecf32` + `vecf64` = `vecf64`
    - `vecf32` + `varchar` = `vecf32`
4. Vector Unary Functions: `summation`, `l1_norm`, `l2_norm`, `vector_dims`
5. Vector Binary Functions: `inner_product`, `cosine_similarity`
6. Disable setting `vector` as `Primary Key`. We also don't support `Unique Index` on `vector` column.
7. Agg Operator : `count`

More details can be found [here](https://github.com/matrixorigin/docs/pull/213).

## NOTE
- Vector type is implemented using `Varlena`.
- Vecf32 is internally called `T_array_float32` (ie `ARRAY`) to avoid getting confused with the Vector class from vectorized execution engine.
- Vector/Array is internally stored as []byte and only converted to []T when there is comparison, toString, or builtin operation.
- Vector/Array is seen by DN as []byte. 
- Zone map of Array type is not very useful. We will be either not initializing this zm or setting theoretic min/max for this zm. 
- Array/Vector index will be done separately via HNSW algo libraries.

## Pending (Maybe in later PR)
- ~ZM handle~
- ~Inf/Nan handle~
- Support Agg operator ANY()
- Support BINARY io 
- Remove auto zero padding and include `pad(arr, value, n)` and `truncate(arr, n)`
- Add UT for `vecf32` in MySQL dialect pkg.
- Add support `infile` export. https://github.com/matrixorigin/matrixone/issues/11446
- Handle errors like b/b with zero padding, panic slice not a multiple of element size, and dimension mismatch comparator.
- Get requirement for compare. [1,2,3] > [1,2] is it this valid?

### CODE REVIEW

**Review by fengttt — COMMENTED** [review]
Test cases, esp agg, still lacks coverage.

**Review by XuPeng-SH — COMMENTED** [review]
As discussed offline, it is necessary to consider the processing of zone map and the test of flushing in the bvt

**Comment by fengttt** [comment]
> Test cases, esp agg, still lacks coverage.

Never mind, count is the only supported agg in this PR.   Need more coverage in operator though -- do we allow mismatch dimensions in operator/functions (0 fill?) ?  Do we handle Inf/Nan?

**Comment by fengttt** [comment]
> As discussed offline, it is necessary to consider the processing of zone map and the test of flushing in the bvt

I assume there is nothing to store in zonemap, other than count and null count?     min/max/sum etc all do not make much sense for vectors.   index is another project.

File: pkg/container/types/array.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+func ArraysToString[T RealNumbers](input [][]T) string {
+	strValues := make([]string, len(input))
+	for i, row := range input {
+		strValues[i] = ArrayToString[T](row)
+	}
+	return strings.Join(strValues, " ")
+}
+
+func StringToArray[T RealNumbers](str string) ([]T, error) {
+	input := strings.TrimSpace(str)
```
Comment (fengttt) [inline pkg/container/types/array.go:64]: isn't this very slow?   first you trim sapce, then check prefix/suffix, then replaceall.   later split string ...

why cannot we use a loop ...

**Comment by XuPeng-SH** [comment]
> 

I also don't think these make sense for vectors. But if the filter conditions of the vector are pushed, we need to deal with it. Of course, it is also possible to write some logic to ignore the zm filtering on the vector. This PR needs to verify that it works properly when ZM is not set processing vector types

File: pkg/vm/engine/tae/index/zm.go [hunk 2]
```diff
@@ -356,6 +357,11 @@ func (zm ZM) getValue(buf []byte) any {
 	case types.T_char, types.T_varchar, types.T_json,
 		types.T_binary, types.T_varbinary, types.T_blob, types.T_text:
 		return buf
+	case types.T_array_float32:
+		// Used by MO_TABLE_COL_MAX
+		return types.BytesToArray[float32](buf)
+	case types.T_array_float64:
+		return types.BytesToArray[float64](buf)
```
Comment (arjunsk) [inline pkg/vm/engine/tae/index/zm.go:364]: Good.

File: pkg/sql/plan/function/func_mo.go [hunk 3]
```diff
@@ -354,6 +354,12 @@ func getValueInStr(value any) string {
 		return v
 	case []byte:
 		return string(v)
+	case []float32:
```
Comment (arjunsk) [inline pkg/sql/plan/function/func_mo.go:357]: OK.

File: pkg/vm/engine/tae/compute/compare.go [hunk 4]
```diff
@@ -158,6 +168,11 @@ func CompareGeneric(a, b any, t types.T) int {
 	case types.T_char, types.T_varchar, types.T_blob,
 		types.T_binary, types.T_varbinary, types.T_json, types.T_text:
 		return CompareBytes(a.([]byte), b.([]byte))
+	case types.T_array_float32:
+		// Used by GetRowByFilter in TAE.
+		return CompareArrayFromBytes[float32](a.([]byte), b.([]byte))
+	case types.T_array_float64:
+		return CompareArrayFromBytes[float64](a.([]byte), b.([]byte))
```
Comment (arjunsk) [inline pkg/vm/engine/tae/compute/compare.go:175]: Nothing to do with ZM.

**Comment by arjunsk** [comment]
> > Test cases, esp agg, still lacks coverage.
> 
> Never mind, count is the only supported agg in this PR. Need more coverage in operator though -- do we allow mismatch dimensions in operator/functions (0 fill?) ? Do we handle Inf/Nan?

@fengttt 
BVT updated.

1. We don't allow mismatched dimensions during operation.
2. We do zero fill during insert if inserted vector is having lesser dimension than its definition.
3. Inf/Nan is not handled. [working on it.]

File: pkg/container/types/array.go (hunk 1 above)
Comment (arjunsk) [reply pkg/container/types/array.go:62]: unsafe access to StringIndex did not yield much of a performance improvement.

```
BenchmarkStringToArray
BenchmarkStringToArray/StringToArray-Trim_Split
BenchmarkStringToArray/StringToArray-Trim_Split-10         	1000000000	         0.0000722 ns/op
BenchmarkStringToArray/StringToArray-ForLoop-RuneCast
BenchmarkStringToArray/StringToArray-ForLoop-RuneCast-10   	    4512	    250405 ns/op
BenchmarkStringToArray/StringToArray-ForLoop-Unsafe
BenchmarkStringToArray/StringToArray-ForLoop-Unsafe-10     	    6218	    187644 ns/op
```

<details>
<summary>StringToArrayV3</summary>

```go


// StringToArrayV3 this implementation uses only one forloop without casting string to rune array (via unsafe)
func StringToArrayV3[T RealNumbers](str string) ([]T, error) {
	strLen := len(str)
	strIdx := 0

	// Trim Space from beginning
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// If beginning is not [ it is malformed.
	if unsafeStringAt(str, strIdx) != '[' {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// skip [
	strIdx++

	// we take the subarray from x based on final array dimension.
	x := make([]T, MaxArrayDimension)
	dim := 0
	var err error

	// find the next "," from the strIdx. More like IndexOf(",", startIdx) from java.
	commaPos := indexFrom(str, ",", strIdx)

	for strIdx < strLen && unsafeStringAt(str, strIdx) != ']' {
		if dim == MaxArrayDimension {
			return nil, moerr.NewInternalErrorNoCtx("typeLen is over the MaxVectorLen: %v", MaxArrayDimension)
		}
		// skip space in the numStr, Eg:- "[  1, 2]"
		for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
			strIdx++
		}

		// if there is no "," then break and check for the "]" condition
		if commaPos == -1 {
			break
		}

		// this means that numStr is ""
		if strIdx == commaPos {
			return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
		}

		// convert numStr to T
		x[dim], err = stringToT[T](str[strIdx:commaPos])
		if err != nil {
			return nil, err
		}
		dim++

		// move strIdx to pos after commaPos
		strIdx = commaPos + 1

		// find the next comma pos from strIdx.
		// If there is none, the break condition kicks in and rest of the code checks for
		// "]" condition.
		commaPos = indexFrom(str, ",", strIdx)
	}

	// skip space for case:- [   ]
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// find next occurrence of "]" from strIdx.
	endBracketPos := indexFrom(str, "]", strIdx)
	if endBracketPos == -1 {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// if numStr is ""
	if strIdx == endBracketPos {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// convert numStr to T and incr dim
	x[dim], err = stringToT[T](str[strIdx:endBracketPos])
	if err != nil {
		return nil, err
	}
	dim++

	// set strIdx after endBracket
	strIdx = endBracketPos + 1

	// Trim of the trailing spaces
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// If after TrimEnd, there are still data in the str, then it is malformed.
	// Eg:- [1,2,3] 1
	if strIdx != strLen {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// return subarray of x based on the final dimension.
	return x[:dim], nil
}

// unsafeStringAt used when we want str[idx] without casting str to []rune
func unsafeStringAt(str string, idx int) rune {
	return rune(*(*byte)(unsafe.Pointer((*reflect.StringHeader)(unsafe.Pointer(&str)).Data + uintptr(idx))))
}

// indexFrom Find the next occurrence of substr after start
func indexFrom(str, substr string, start int) int {
	pos := strings.Index(str[start:], substr)
	if pos != -1 {
		pos += start
	}
	return pos
}

// stringToT convert str to T
func stringToT[T RealNumbers](str string) (t T, err error) {
	switch any(t).(type) {
	case float32:
		num, err := strconv.ParseFloat(str, 32)
		if err != nil {
			return t, moerr.NewInternalErrorNoCtx("error while casting %s to %s", str, T_float32.String())
		}
		// FIX: https://stackoverflow.com/a/36391858/1609570
		numf32 := float32(num)
		return *(*T)(unsafe.Pointer(&numf32)), nil
	case float64:
		num, err := strconv.ParseFloat(str, 64)
		if err != nil {
			return t, moerr.NewInternalErrorNoCtx("error while casting %s to %s", str, T_float64.String())
		}
		return *(*T)(unsafe.Pointer(&num)), nil
	default:
		panic(moerr.NewInternalErrorNoCtx("not implemented"))
	}
}


```
</details>

Comment (arjunsk) [reply pkg/container/types/array.go:62]: @fengttt 
I think we should need [Binary Insert](https://github.com/matrixorigin/matrixone/issues/9955) for better performance.

Based on a quick benchmark, we could get up to 8x performance
```
    # mo       macos   run2: Result: vector dim=1024 vectors inserted=40960 insert/second=2650.0632385796366 Binary
    # mo       macos   run3: Result: vector dim=1024 vectors inserted=40960 insert/second=2701.284438140351 BinaryString
```

**Review by fengttt — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by xzxiong — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by triump2020 — APPROVED** [review]

File: pkg/frontend/export.go [hunk 5]
```diff
@@ -465,6 +465,7 @@ func constructByte(obj interface{}, bat *batch.Batch, index int32, ByteChan chan
 			case types.T_char, types.T_varchar, types.T_blob, types.T_text, types.T_binary, types.T_varbinary:
 				value := addEscapeToString(vec.GetBytesAt(i))
 				writeByte = appendBytes(writeByte, value, symbol[j], closeby, true)
+				//TODO: How should we export T_array in CSV? Is it as "[1,2,3]"?
```
Comment (daviszhen) [inline pkg/frontend/export.go:468]: can the vector type be exported to string? missing here

**Review by qingxinhome — APPROVED** [review]

File: pkg/frontend/export.go (hunk 5 above)
Comment (arjunsk) [reply pkg/frontend/export.go:468]: @flying-x could you kindly share your thoughts on this?

**Review by badboynt1 — APPROVED** [review]

File: test/distributed/cases/array/array.sql [hunk 6]
```diff
@@ -0,0 +1,80 @@
+-- pre
+drop database if exists vecdb;
+create database vecdb;
```
Comment (daviszhen) [inline test/distributed/cases/array/array.sql:3]: add some cases like "SELECT * into outfile 'a.txt'" using the vector type

**Review by nnsgmsone — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

File: test/distributed/cases/array/array.sql (hunk 6 above)
Comment (arjunsk) [reply test/distributed/cases/array/array.sql:3]: https://github.com/matrixorigin/matrixone/issues/11446

File: pkg/sql/colexec/s3util.go [hunk 7]
```diff
@@ -475,6 +475,7 @@ func (w *S3Writer) SortAndFlush(proc *process.Process) error {
 			merge = NewMerge(len(w.Bats), sort.NewUuidCompLess(), getFixedCols[types.Uuid](w.Bats, pos), nulls)
 		case types.T_char, types.T_varchar, types.T_blob, types.T_text:
 			merge = NewMerge(len(w.Bats), sort.NewGenericCompLess[string](), getStrCols(w.Bats, pos), nulls)
+			//TODO: check if we need T_array here? T_json is missing here.
```
Comment (arjunsk) [inline pkg/sql/colexec/s3util.go:478]: Handle it in https://github.com/matrixorigin/matrixone/issues/11446

File: pkg/sql/util/eval_expr_util.go [hunk 8]
```diff
@@ -601,12 +614,62 @@ func setInsertValueString(proc *process.Process, numVal *tree.NumVal, vec *vecto
...
+				if len(_v) > destLen {
+					return nil, function.FormatCastErrorForInsertValue(proc.Ctx, s, *typ, fmt.Sprintf("Src length %v is larger than Dest length %v", len(_v), destLen))
+				}
+			case types.T_array_float64:
+				_v, err := types.StringToArray[float64](s)
+				if err != nil {
+					return nil, err
+				}
+
+				// zero padding from []T{1,2} to T{1,2,0}
```
Comment (arjunsk) [inline pkg/sql/util/eval_expr_util.go:651]: Remove zero padding and throw an error if dimension mismatch.

File: pkg/sql/util/index_util.go [hunk 9]
```diff
@@ -284,11 +284,14 @@ func serialWithCompacted(vs []*vector.Vector, proc *process.Process) (*vector.Ve
 				}
 			}
 		case types.T_json, types.T_char, types.T_varchar, types.T_binary, types.T_varbinary, types.T_blob, types.T_text:
+			//TODO: Is T_array required here?
 			vs := vector.MustStrCol(v)
 			for i := range vs {
 				if nulls.Contains(v.GetNulls(), uint64(i)) {
 					nulls.Add(bitMap, uint64(i))
 				} else {
+
```
Comment (arjunsk) [inline pkg/sql/util/index_util.go:293]: Something pending.

File: pkg/util/export/etl/tae.go [hunk 10]
```diff
@@ -216,6 +217,7 @@ func getOneRowData(ctx context.Context, bat *batch.Batch, Line []table.ColumnFie
 			cols[rowIdx] = field.GetFloat64()
 		case types.T_char, types.T_varchar,
 			types.T_binary, types.T_varbinary, types.T_blob, types.T_text:
+			//TODO: How to handle T_array here?
```
Comment (arjunsk) [inline pkg/util/export/etl/tae.go:220]: To handled in https://github.com/matrixorigin/matrixone/issues/11446

File: pkg/vm/engine/tae/common/print.go [hunk 11]
```diff
@@ -124,6 +124,11 @@ func TypeStringValue(t types.Type, v any, isNull bool, opts ...TypePrintOpt) str
 		} else {
 			return fmt.Sprintf("%x", buf)
 		}
+	case types.T_array_float32:
```
Comment (arjunsk) [inline pkg/vm/engine/tae/common/print.go:127]: Handled in https://github.com/matrixorigin/matrixone/issues/11446

File: pkg/sql/util/eval_expr_util.go (hunk 8 above)
Comment (arjunsk) [reply pkg/sql/util/eval_expr_util.go:651]: Handled in this https://github.com/matrixorigin/matrixone/issues/11458

File: pkg/sql/util/index_util.go [hunk 12]
```diff
@@ -284,11 +284,14 @@ func serialWithCompacted(vs []*vector.Vector, proc *process.Process) (*vector.Ve
...
+					// TODO: If T_array is required, we might need to define arrayTypeCode in tuples.
 					ps[i].EncodeStringType([]byte(vs[i]))
```
Comment (arjunsk) [inline pkg/sql/util/index_util.go:295]: Hi @qingxinhome could you kindly share your feedback on this?

Comment (arjunsk) [reply pkg/sql/util/index_util.go:295]: Could you please check if this https://github.com/arjunsk/matrixone/pull/42 is good enough?

Comment (qingxinhome) [reply pkg/sql/util/index_util.go:295]: I agree with your view that VECTOR and ARRAY can be limited to not being primary or unique keys. In fact, in Hive, VECTOR and ARRAY types cannot be directly used as primary keys. In PostgreSQL, VECTOR and ARRAY types can be used as part of the primary key, but cannot be directly used as the entire primary key.But you can listen to the product manager's @domingozhang opinion

**Pull request merged**